mixer/pkg/server/interceptor.go

```diff
@@ -0,0 +1,106 @@
+package server
+
+import (
+	"context"
+	"strings"
+
+	"github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc"
+	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+	"github.com/opentracing/opentracing-go/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	// Morally a const:
+	componentTag = opentracing.Tag{Key: string(ext.Component), Value: "istio-mixer"}
+
+	defaultNoopTracer = opentracing.NoopTracer{}
+)
+
+type metadataReaderWriter struct {
+	metadata.MD
+}
+
+func (w metadataReaderWriter) Set(key, val string) {
+	// The GRPC HPACK implementation rejects any uppercase keys here.
+	//
+	// As such, since the HTTP_HEADERS format is case-insensitive anyway, we
+	// blindly lowercase the key (which is guaranteed to work in the
+	// Inject/Extract sense per the OpenTracing spec).
+	key = strings.ToLower(key)
+	w.MD[key] = append(w.MD[key], val)
+}
+
+func (w metadataReaderWriter) ForeachKey(handler func(key, val string) error) error {
+	for k, vals := range w.MD {
+		for _, v := range vals {
+			if err := handler(k, v); err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// TracingServerInterceptor copy of the GRPC tracing interceptor to enable switching
+// between the supplied tracer and NoopTracer depending upon whether the request is
+// being sampled.
+func TracingServerInterceptor(tracer opentracing.Tracer) grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (resp interface{}, err error) {
+		md, ok := metadata.FromIncomingContext(ctx)
+		if !ok {
+			md = metadata.New(nil)
+		}
+
+		var t opentracing.Tracer = defaultNoopTracer
+		if isSampled(md) {
+			t = tracer
+		}
+
+		spanContext, err := t.Extract(opentracing.HTTPHeaders, metadataReaderWriter{md})
```

Comment:
in the other PR, we were checking to see if this `err` was `opentracing.ErrSpanContextNotFound`. This, I think, was forcing noopSpans for cases when no trace headers were forwarded (which I believe is still the desired behavior). Should something like that be included here?